Frameworks/CoreGraphics/CGColor.mm

```diff
@@ -14,73 +14,156 @@
+            0, // Version
+            typeid(__CGColor).name(), // Class Name
+            nullptr, // Init
+            nullptr, // Copy (ill-defined for CF)
+            &_CFFinalize, // Finalize
+            &__CFEquals, // Equals
+            nullptr, // Hash
+            nullptr, // Copy Description
+            nullptr, // Copy Debug Description
+        };
+        static CFTypeID _id = _CFRuntimeRegisterClass(&_cls);
+        return _id;
+    }
+
+    __CGColor(CGColorSpaceRef colorspace, CGFloat r, CGFloat g, CGFloat b, CGFloat alpha)
+        : _colorSpace(colorspace), _pattern(nullptr), _components{ r, g, b, alpha } {
+    }
+
+    __CGColor(CGColorSpaceRef colorSpace, CGPatternRef pattern) : _colorSpace(colorSpace), _pattern(pattern) {
+        // TODO #2581: CGColorCreateWithPattern should support components
+    }
+
+    // CreateCopyWithAlpha
+    __CGColor(const __CGColor& color, float alpha) : _colorSpace(color._colorSpace), _pattern(color._pattern) {
+        if (color.IsPatternColor()) {
+            _components.Clear();
+            _components.a = alpha;
+        } else {
+            _components.SetColorComponents(color._components.r, color._components.g, color._components.b, alpha);
+        }
+    }
+
+    bool operator==(const __CGColor& other) const {
+        return (CGColorSpaceGetModel(_colorSpace) == CGColorSpaceGetModel(other._colorSpace)) && (_pattern == other._pattern) &&
+               (_components == other._components);
+    }
+
+    inline CGFloat Alpha() const {
+        return _components.a;
+    }
+
+    inline CGColorSpaceRef ColorSpace() const {
+        return _colorSpace;
+    }
+
+    inline const __CGColorQuad& ColorComponents() const {
+        return _components;
+    }
+
+    inline CGPatternRef Pattern() const {
+        return _pattern;
+    }
+
+private:
+    __CGColorQuad _components;
+    woc::StrongCF<CGColorSpaceRef> _colorSpace;
+    woc::StrongCF<CGPatternRef> _pattern;
+
+    static Boolean __CFEquals(CFTypeRef color1, CFTypeRef color2) {
+        return *((__CGColor*)color1) == *((__CGColor*)color2);
+    }
+
+    inline bool IsPatternColor() const {
+        return _pattern != nullptr;
+    }
+};
+
+/**
+ @Status Caveat
+ @Notes components are not supported
+*/
+CGColorRef CGColorCreateWithPattern(CGColorSpaceRef colorSpace, CGPatternRef pattern, const CGFloat* components) {
+    RETURN_NULL_IF(!colorSpace);
+    RETURN_NULL_IF(!pattern);
+    return __CGColor::CreateInstance(kCFAllocatorDefault, colorSpace, pattern);
+}
+
 /**
  @Status Interoperable
 */
 CGColorRef CGColorGetConstantColor(CFStringRef name) {
-    CGColorRef ret = nullptr;
-    woc::StrongCF<CGColorSpaceRef> colorspace{ woc::TakeOwnership, CGColorSpaceCreateDeviceGray() };
     if (CFEqual(kCGColorBlack, name)) {
-        ret = static_cast<CGColorRef>([[__LazyUIColor blackColor] CGColor]);
+        static woc::StrongCF<CGColorRef> s_black{ woc::TakeOwnership, CGColorCreateGenericGray(0.0f, 1.0f) };
```

Comment:
grayspace 

---
In reply to: [113502948](https://github.com/Microsoft/WinObjC/pull/2538#discussion_r113502948) [](ancestors = 113502948)